test/e2e/network/netpol/model.go

```diff
@@ -0,0 +1,286 @@
+			return cont, nil
+		}
+	}
+	return nil, errors.Errorf("unable to find container in pod %s/%s, port %d, protocol %s", p.Namespace, p.Name, port, protocol)
+}
+
+// PodString returns a corresponding pod string
+func (p *Pod) PodString() PodString {
+	return NewPodString(p.Namespace, p.Name)
+}
+
+// ContainerSpecs builds kubernetes container specs for the pod
+func (p *Pod) ContainerSpecs() []v1.Container {
+	var containers []v1.Container
+	for _, cont := range p.Containers {
+		containers = append(containers, cont.Spec())
+	}
+	return containers
+}
+
+func (p *Pod) labelSelectorKey() string {
+	return "pod"
+}
+
+func (p *Pod) labelSelectorValue() string {
+	return p.Name
+}
+
+// LabelSelector returns the default labels that should be placed on a pod/deployment
+// in order for it to be uniquely selectable by label selectors
+func (p *Pod) LabelSelector() map[string]string {
+	return map[string]string{
+		p.labelSelectorKey(): p.labelSelectorValue(),
+	}
+}
+
+// KubePod returns the kube pod
+func (p *Pod) KubePod() *v1.Pod {
+	zero := int64(0)
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      p.Name,
+			Labels:    p.LabelSelector(),
+			Namespace: p.Namespace,
+		},
+		Spec: v1.PodSpec{
+			TerminationGracePeriodSeconds: &zero,
+			Containers:                    p.ContainerSpecs(),
+		},
+	}
+}
+
+// QualifiedServiceAddress returns the address that can be used to hit a service from
+// any namespace in the cluster
+func (p *Pod) QualifiedServiceAddress() string {
+	return fmt.Sprintf("%s.%s.svc.cluster.local", p.ServiceName(), p.Namespace)
+}
+
+// ServiceName returns the unqualified service name
+func (p *Pod) ServiceName() string {
+	return fmt.Sprintf("s-%s-%s", p.Namespace, p.Name)
+}
+
+// Service returns a kube service spec
+func (p *Pod) Service() *v1.Service {
+	service := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      p.ServiceName(),
+			Namespace: p.Namespace,
+		},
+		Spec: v1.ServiceSpec{
+			Selector: p.LabelSelector(),
+		},
+	}
+	for _, container := range p.Containers {
+		service.Spec.Ports = append(service.Spec.Ports, v1.ServicePort{
+			Name:     fmt.Sprintf("service-port-%s-%d", strings.ToLower(string(container.Protocol)), container.Port),
+			Protocol: container.Protocol,
+			Port:     container.Port,
+		})
+	}
+	return service
+}
+
+// Container is the abstract representation of what matters to network policy tests for
+// a container; i.e. it ignores kube implementation details
+type Container struct {
+	Port     int32
+	Protocol v1.Protocol
+}
+
+// Name returns the container name
+func (c *Container) Name() string {
+	return fmt.Sprintf("cont-%d-%s", c.Port, strings.ToLower(string(c.Protocol)))
+}
+
+// PortName returns the container port name
+func (c *Container) PortName() string {
+	return fmt.Sprintf("serve-%d-%s", c.Port, strings.ToLower(string(c.Protocol)))
+}
+
+// Spec returns the kube container spec
+func (c *Container) Spec() v1.Container {
```

Comment:
question, does it makes sense have a pod listening in multiple protocols, i.e. tcp and udp ?

Reply:
@eth0XFEED %^

Reply:
@aojea i think we have the pod listening for both protocols now but its being accomplished in a multiple container per pod strategy. 

Do i understand correct that you would like to see single container in pod which is binding to both types of protocol sockets ?

Reply:
No no no, no suggesting anything, just that iirc agnhost supports it without multiple containers, is just adding more flags to the agnhost command ... Was wondering if you were aware of this

Reply:
I think the separate containers for different port affords us more flexibility long term, but I could see this going either way